bytecomp/bytegen.ml

```diff
@@ -897,19 +897,29 @@ let rec comp_expr env exp sz cont =
           let c = comp_expr env lam sz cont in
           let ev = event Event_pseudo Event_other in
           add_event ev c
-      | Lev_after _ when is_tailcall cont -> (* don't destroy tail call opt *)
-          comp_expr env lam sz cont
       | Lev_after ty ->
-          let info =
+          let preserve_tailcall =
             match lam with
-              Lapply{ap_args = args}  -> Event_return (List.length args)
-            | Lsend(_, _, _, args, _) -> Event_return (List.length args + 1)
-            | Lprim(_,args,_)         -> Event_return (List.length args)
-            | _                       -> Event_other
+            | Lprim((Pidentity | Popaque | Pdirapply | Prevapply), _, _) ->
+               true
+            | Lprim(_, _, _) -> false
```

Comment:
I understand that it is a pain, but I would be more confident with an exhaustive listing of primitives rather than a fragile pattern-matching here, so that people adding new primitives have to consider this question. (Correspondingly, I would expect a comment explaining what this check is supposed to do, and I think it would be nicer to have it as a lam-parametrized toplevel function.)

Reply:
Sorry, but this is *really* a pain: this algebraic type has more than 100 constructors!

I added a comment. As for new primitives: I would argue that it is very unlikely that such new primitives would require a change here. It is actually already arguable whether `f x` is a tail call in e.g., `Obj.magic (f x)` when this is in tail position.

Reply:
To finish reviewing and approve this PR, I feel that I should check disabling tailcalls for all the primitives not listed above is correct. I guess that this involves looking at how each of them is compiled. Having a quick look at the list of primitives, I wonder about `Pignore, Praise, Psequand, Psequor, Pint_as_pointer`; could those not result in tailcalls in some situations?

Reply:
`Praise` : this is compiled into the `Kraise` instruction, which never returns. The question of preserving the tail continuation is irrelevant.

`Pint_as_pointer`: this is compiled into a C call to the `caml_int_as_pointer` function

`Psequand`, `Psequor`: you're right. Fixed. (Although this seems to neverresult in a bug since an "event after" is never inserted around these primitives.)

Reply:
Thinking about this more, I think that it's one of these situations where we would regret not having written a complete pattern-matching, after the fact. I would like to list all primitives, and I think that it could be kept readable by moving it as a toplevel function (like `is_tailcall`) instead of being in the middle of the code. I'm happy to push a new commit to do this myself if you don't have the time.

Reply:
Alright, let me do that.